### PR TITLE
[MRG] Add probability arguments to drives API

### LINKED ISSUE #410: warning for no connections in Network object
moving @ntolley 's comment to separate issue:

https://github.com/jonescompneurolab/hnn-core/pull/77#discussion_r683714765

raise a warning of `net.connectivity` has no local connections (but input drives?)

### PR DESCRIPTION
**Pull request by ntolley**
Closes #410. Thanks to #369, this task is quite straightforward. The goal of this PR is to add a probability argument to the drives API, such that `net.add_evoked_drive(..., probability=0.5)` randomly removes half the drive connections, and only a subset are actually targeted by the drive. Additionally, the probability argument should also accept a dictionary which specifies probabilities for each individual connection (same was the weight and delay arguments).

After completing this, I think the final task for fully merging the drives and connection API's is enabling the manual targeting of `target_gids`, but let's leave that as a topic of discussion for the upcoming retreat ;).

### CODE REVIEW

**Comment by ntolley** [comment]
Also I want to mention a potential concern I had that ended up not being a problem. `net.connectivity` contains a separate element for each receptor type, there is the potential that a given `src_gid -> target_gid` pair may be missing the full suite of receptors intended for that connection. For example, the GABAa and GABAb connections from the same inhibitory `src_gid` targeting different neurons.

Rather conveniently, this is gracefully handled by making sure the same seed is used when calling `net.add_connection` for the receptors contained in the same "connection class". I'll add some tests to check this explicitly, but by manual inspection the src-target pairs agree across receptors for the same connection class.

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:607]: why not reuse seedcore?

Comment (ntolley) [reply hnn_core/network.py:607]: Personally I think this is pretty important to decouple the random seeding of the events vs connections. I can certainly imagine a case for wanting to keep the same physical connectivity structure while changing the event times.

Perhaps rename `seedcore` to `event_seed`?

Comment (jasmainak) [reply hnn_core/network.py:607]: okay, yeah we can call it `event_seed`. By the way, we are doing this kind of renaming very liberally right now but in a more stable software it would break user scripts. I think we can continue to do this until 0.2 but start getting into the habit of deprecation warnings after that.

Also, the more I think, the more I like your proposal of making `Drives` into an object like cells so you can just use the network connectivity API.

File: hnn_core/dipole.py
Comment (jasmainak) [inline hnn_core/dipole.py:58]: use `len` because `net.connectivity` is list not boolean

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:826]:
```suggestion
                        probability=probability, seed=drive['conn_seed'])
```

I think here it's obvious that the seed is related to connectivity

**Comment by jasmainak** [comment]
add a test?

**Comment by jasmainak** [comment]
Also, can you add closes #410 in the description of the PR ?

File: hnn_core/dipole.py
Comment (ntolley) [reply hnn_core/dipole.py:58]: You've unknowingly stumbled into the most controversial opinion the 21st century :smile: 
https://stackoverflow.com/questions/53513/how-do-i-check-if-a-list-is-empty
https://stackoverflow.com/questions/6459513/why-implicitly-check-for-emptiness-in-python

Comment (jasmainak) [reply hnn_core/dipole.py:58]: interesting Python feature :) I think I still prefer the explicit form though but you can decide

File: hnn_core/network.py
Comment (ntolley) [reply hnn_core/network.py:837]: Totally agree, but I prefer to have named arguments stay the same when passed to multiple functions. Not a big deal to change if you disagree for this situation though.

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:574]: what is delay->cell mapping?

File: hnn_core/tests/test_drives.py
Comment (jasmainak) [inline hnn_core/tests/test_drives.py:362]: I realize I have myself done this in the past. But I guess it might be useful to spell out the full message here because I can't distinguish between this test and the "probability is either" one without searching in the codebase for that phrase

**Comment by jasmainak** [comment]
Few minor comments otherwise LGTM

I see that there are a couple of other todos in the release checklist but this was not amongst them. Would you mind updating that checklist to reflect the latest shopping list?

**Comment by ntolley** [comment]
Can do! And apologies going off script w.r.t. release priorities, this came up as being very necessary for some preliminary grant analysis.

File: hnn_core/network.py
Comment (ntolley) [reply hnn_core/network.py:574]: Oops, typo. This was copied from the delay parameter for the drives API, which has identical data structures for input

**Comment by jasmainak** [comment]
@ntolley looks like we have a CI failure due to MNE. Can you reproduce the import issue locally (with a Python 3.6 conda env and doing pip install mne)? If so, we should report it to MNE, otherwise we can just restart the CI

File: hnn_core/tests/test_drives.py
Comment (ntolley) [reply hnn_core/tests/test_drives.py:178]: Following up from our previous conversation on this, it seems like this reduced connection may not be very practical since receptors would be defined with different weights as well. 

Also @rythorpe, all drive connections created from a single call share the same seed. So when the connection gets passed to `net.add_connection` with probability <1.0, the same connections will be preserved across receptors.

Comment (jasmainak) [reply hnn_core/tests/test_drives.py:178]: sounds good. I'm a little uneasy with ensuring "sameness" with seeds. I worry that we'll get weird bugs. 

Alternative is to allow `add_connection` to accept list of receptors and then have [`connection_probability`](https://github.com/jonescompneurolab/hnn-core/blob/20b6903b0682473f2af20357e086e1a980f26712/hnn_core/network.py#L97) return the `conn_pairs` to drop, then drop those `conn_pairs` for both receptors. I realize it's a bit of work, so let's move on here. But worth keeping in mind if such an issue comes up in the future

**Review by jasmainak — APPROVED** [review]
merge button is yours @rythorpe

File: hnn_core/tests/test_drives.py
Comment (ntolley) [reply hnn_core/tests/test_drives.py:193]: Just want to submit a few DOC updates in a commit, please feel free to merge afterwards

Comment (ntolley) [reply hnn_core/tests/test_drives.py:195]: Good to go on my end once the CI's are green!

File: hnn_core/tests/test_drives.py
Comment (ntolley) [reply hnn_core/tests/test_drives.py:178]: Very valid, moving to a separate issue

Comment (rythorpe) [reply hnn_core/tests/test_drives.py:178]: It looks like `conn_seed` enforces the same "random" connection pattern across cell types as well. Thus, "random" connectivity is not statistically independent of other cell types (e.g., specifying 60% connection probability between a proximal evoked drive and L2/3 pyramidal cells will give the exact same connectivity pattern as for a 60% connection probability to L5 pyramidal cells). This seems sketchy to me. As a user generally speaking, I would by default assume that any pseudo-randomized process is free of structural correlations unless otherwise stated.

Comment (ntolley) [reply hnn_core/tests/test_drives.py:178]: This is a really good point, I'm putting this back to WIP until I add a cleaner way to handle this case.

File: hnn_core/network.py
Comment (ntolley) [inline hnn_core/network.py:844]: @rythorpe I think this addresses your problem with enforcing shared connections with seeding. The connections for the same receptor are explicitly set to be equal to one another.

I additionally added `seed_increment` to make sure that there aren't weird structural dependencies for the random subsets. I see a PR in the future that takes care of all the seeding as we discussed previously (seeds drawn from a single RNG).

File: hnn_core/tests/test_dipole.py
Comment (rythorpe) [inline hnn_core/tests/test_dipole.py:138]: Nitpick: which line actually returns the warning? If it is both instances of `simulate_dipole()`, should this be broken into 2 context managers that each raise the same warning?

File: hnn_core/network.py
Comment (rythorpe) [inline hnn_core/network.py:97]: Out of curiosity, why did you need to change `conn_seed` to 3?

File: hnn_core/network.py
Comment (rythorpe) [inline hnn_core/network.py:784]: If the user doesn't specify the connection probability for a certain cell type, should the default legacy mode behavior be to set the probability at 1.0 or 0?

File: hnn_core/network.py
Comment (rythorpe) [inline hnn_core/network.py:847]: Is this necessary given that `conn_seed` applies the same `seed_increment` for each receptor of a given cell type?

**Comment by jasmainak** [comment]
@ntolley you need to rebase

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:500]: I think there is some missing documentation here. Normally seed is `None` by default. Here you set it an integer so that the connections are consistent across trials? Would be good to spell out across what the connections are consistent

File: hnn_core/network.py
Comment (jasmainak) [reply hnn_core/network.py:844]: perhaps raise an issue with crosslink here to not forget?

**Comment by jasmainak** [comment]
@ntolley don't forget to update `whats_new.rst` !

File: hnn_core/network.py
Comment (ntolley) [reply hnn_core/network.py:847]: This is actually the root of the problem we were discussing with respect to enforcing matching connections. While technically these would share the same connections when providing the same seed, it's not very transparent, so they are explicitly set equal here.

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:858]:
```suggestion
                    # Ensure that AMPA/NMDA connections target the same gids
                    # when probability < 1
```

Comment (rythorpe) [reply hnn_core/network.py:858]: If I'm not mistaken, this enforces similar AMPA and NMDA connections regardless of the probability.

File: hnn_core/network.py
Comment (ntolley) [reply hnn_core/network.py:102]: Honestly I don't remember :smile:, perhaps it's a good reminder that the seed can be any int.

File: hnn_core/network.py
Comment (ntolley) [reply hnn_core/network.py:535]: Very good point, I honestly didn't have a motivation behind fixing the seed across trials but it is a pleasant (on unpleasant) side effect depending on the question you're interested in, I'll add a better description

File: hnn_core/network.py
Comment (ntolley) [reply hnn_core/network.py:837]: I think it should default to 1.0 since that is the default when you don't specify `probability` at all

File: hnn_core/network.py
Comment (ntolley) [reply hnn_core/network.py:925]: This is true, but I guess it's worth emphasizing that this is only important when probability < 1.0

**Comment by jasmainak** [comment]
@rythorpe I'm not available until Sunday due to coding sprint in another project. Merge button is yours!

File: hnn_core/network.py
Comment (rythorpe) [reply hnn_core/network.py:914]: Hmm, I think the root problem is addressed by explicitly implementing (with documentation) that random connections are assigned on a cell-by-cell basis rather than a synapse-by-synapse basis. This is perfectly implemented by setting the seed increment for each cell type in an outer loop and then establishing each receptor's connection in an inner loop. That said, maybe the explicit check is good for maintaining this established convention in the future....

**Comment by rythorpe** [comment]
I'll definitely be using this feature in the near future. Thanks @ntolley!!

**Pull request merged**
